Low level kymotracker API (#46)
lumicks/pylake: docs/tutorial/kymotracking.rst

Kymotracking
============

.. only:: html

    :nbexport:`Download this page as a Jupyter notebook <self>`

Kymographs are a vital tool in understanding binding events on DNA. They are generated using one-dimensional scans
over time. Moving particles can then be seen moving along the spatial coordinate in time. It is often desired to
to track the position of such particles in time. Pylake comes with a few features that enable you to do so.

Performing the tracking
-----------------------

Performing Kymotracking is usually performed in two steps. Typically, there is an image processing step, following by
a tracing step, where the actual traces are found.

Greedy Algorithm
----------------

The first method implemented for performing such a tracking is based on :cite:`sbalzarini2005feature,mangeol2016kymographclear`.
It starts by performing peak detection, performing a grey dilation on the image, and detection which pixels remain
unchanged. Peaks that fall below a certain intensity threshold are discarded. Since this peak detection operates at a
pixel granularity, it is followed up by a refinement step to attain subpixel accuracy. This refinement is performed by

> computing an offset from the brightness weighted centroid in a small neighborhood `w` around the pixel.

Author: 👍

.. math::

    offset = \frac{1}{m} \sum_{i^2 < w^2} i I(x + i)

Where m is given by:

.. math::

    m = \sum_{i^2 < w^2} I(x + i)

After peak detection the feature points are linked together using a forward search analogous to

> :cite:`mangeol2016kymographclear`. This in contrast with the linking algorithm in :cite:`sbalzarini2005feature`

Author: 👍

> which uses a graph based optimization approach. This linking step traverses the kymograph, tracing particles starting

Author: 👍

- Once there are no more peaks in the frame from which we are currently initiating traces, we start initiating traces from the next frame. This process is continued until there are no more peaks left to trace.

The score function is based on a prediction of where we expect future peaks. Based on the peak location of the tip of
the trace `x` and a velocity `v`, it computes a predicted position over time. The score function assumes a Gaussian
uncertainty around that prediction, placing the mean of that uncertainty on the predicted extrapolation. The width of
this uncertainty is given by a base width (provided as sigma) and a growing uncertainty over time given by a diffusion
rate. This results in the following model for the connection score.

.. math::

    S(x, t) = N\left(x + v t, \sigma_{base} + \sigma_{diffusion} \sqrt{t}\right).

Here `N` refers to a normal distribution. In addition to the model, we also have to set a cutoff, after which we deem
peaks to be so unlikely to be connected that they shouldn't be. By default, this cutoff is set at two sigma. Scores
outside this cutoff are set to zero which means they will not be accepted as a new point.

Using the greedy algorithm

> --------------------------

Reviewer: The algorithm explanation is excellent. I enjoyed reading it and it's great to have it layed out here for users. But I would flip the order of 1 Details 2 Usage to 1 Usage 2 Details. Most people reading this will come here primarily for the usage explanation and most of the time people don't need to understand how something works, just how to use it. So let's prioritize the usage explanation on top.

Author: Agreed!

So, time to use this algorithm. First, we need to get an image to perform the tracing on. Let's grab a kymograph from
a Bluelake file::

    file = lk.File('kymograph.h5')
    _, kymo = file.kymos.popitem()

We can extract the image data with one of the `_image` commands::

    data = kymo.green_image

Let's have a look at what it looks like::

    plt.imshow(data)

> It seems that our traces in this image are about 5 pixels wide, so let's set this as our `line_width`. The peaks have an

Reviewer: Should there be an image included here?

intensity of about 5 photon counts, whereas the background fluctuates around 2-3. Let's set our `pixel_threshold` to 4.
We also see that sometimes, a particle momentarily disappears. To still connect these in a single trace, we want to
allow for some gaps in the connection step. Let's use a `window` size of 4 in this test. Running the algorithm is easy::

>     traces = track_particles_greedy(data, line_width=5, pixel_threshold=4, window=4)

Reviewer: Needs `lk.` to indicate that it's from `pylake`:
```suggestion
    traces = lk.track_particles_greedy(data, line_width=5, pixel_threshold=4, window=4)
```

Author: 👍

The result is a set of kymograph traces. We can get the determined time and coordinates of these traces from the
`time` and `coordinate` attributes. Let's plot the first trace::

>     plt.plot(traces[0].time, traces[0].coordinate)

Reviewer: Might be good to show something like this first to explain that `traces` is just an ordinary list:
```py
print(len(traces))  # the number of traces found in the kymo
```

Plotting all the lines is also quite easy using a Python list comprehension::

>     [plt.plot(trace.time, trace.coordinate for trace in traces]

Reviewer: As much as I like list comprehensions, I'd go for a regular `for` loop here since we're not actually using the results and this may look a bit strange to novice Python users.

Author: 😁 

Done!

Note that the determined time and coordinates of the traces and coordinates are defined in pixels and therefore need to

> be converted to actual time and position using the delta time and pixel size of the kymograph. We can get the time

Reviewer: I missed this detail during my review of the previous PR. Could we maybe rename the properties `.time` and `.coordiante` to `.time_idx` and `.coordinate_idx` to better reflect this? I think this is an easy mistake otherwise.

Author: Yeah, I considered that as well already but was on the fence about it. But will do.

between frames directly from the kymograph::

>     dt = (kymo.timestamps[0, 1] - kymo.timestamps[0, 0])/1e9

Reviewer: It would be nice to add a specific API to `Kymo` if this will be used frequently.

Author: Added + tests.

And the pixel size as well::

>     pixel_to_um = kymo.json["scan volume"]["scan axes"][0]["pixel size (nm)"] / 1000

Reviewer: Same as above. Even more important since this is pretty obscure and accesses the raw JSON.

Author: Added + tests. For `scan` I'll throw a `NotImplementedError` for now, because I'm not sure whether we should return a list there or just the line time.

This allows conversion from pixels back to position.

Extracting summed intensities
-----------------------------

Sometimes, it can be desirable to extract pixel intensities in a region around our kymograph trace. We can quite easily
extract these using the method `sample_from_image`. For instance, if we want to sum the pixels in a 9 pixel area around
our first kymograph trace, we can invoke::

    trace = traces[0]
    trace.sample_from_image(data, n_pixels=4)

Here `n_pixels` is the number of pixels to sum on either side of the trace. Note that data has to be the same image

> that the trace was based on.

Reviewer: Can we give each trace a reference to the original `data`? That way users don't need to keep track to match the correct `trace` and `data`. There shouldn't be any extra overhead since the data exists anyway, we just have an extra reference.

Differential geometry algorithm
-------------------------------

> The second algorithm is an algorithm that looks for curvilinear structures in an image. This method is based on sections

Reviewer: The docs call this "differential geometry" while the code calls it `lines`. It would be good to reconcile that.

Author: Renamed it to the actual function name and made it a sub sub header under the new section `Algorithms`

    This is early access alpha functionality. While usable, this has not yet been tested in a large number of different
    scenarios. The API is still be subject to change *without any prior deprecation notice*! If you use this
    functionality keep a close eye on the changelog for any changes that may affect your analysis.

Kymotracking
============

.. only:: html

    :nbexport:`Download this page as a Jupyter notebook <self>`

Kymographs are a vital tool in understanding binding events on DNA. They are generated using one-dimensional scans
over time. Moving particles can then be seen moving along the spatial coordinate in time. It is often desired to
to track the position of such particles in time. Pylake comes with a few features that enable you to do so.

Performing the tracking
-----------------------

Performing Kymotracking is usually performed in two steps. Typically, there is an image processing step, following by
a tracing step, where the actual traces are found.

Using the greedy algorithm
--------------------------

First, we need to get an image to perform the tracing on. Let's grab a kymograph from a Bluelake file::

    file = lk.File('kymograph.h5')

>     _, kymo = file.kymos.popitem()

Reviewer: Keep the name here even if we're not using it. This could be someone's first time seeing this pattern.
```suggestion
    name, kymo = file.kymos.popitem()
```

Author: Good point.

We can extract the image data with one of the `_image` commands::

    data = kymo.green_image

Let's have a look at what it looks like::

    plt.imshow(data, aspect="auto", vmax=5)

.. image:: kymo_unscaled.png

What we also see is that we can clearly see the bead edges in the kymograph. Let's only take a subsection of the
kymograph, to make sure that we don't get any spurious line detections that come from the bead edges. We only use the
70th to the 220th pixel.

    >>> data = data[70:220, :]

If we zoom in a bit, we can see that our traces in this image are about 3 pixels wide, so let's set this as our
`line_width`.

.. image:: kymo_zoom.png

> The peaks have an intensity of about 3-7 photon counts, whereas the background fluctuates around 0-2. Let's set our

Reviewer: Consider adding a colorbar to the plot to make this more obvious in the docs without interactive widgets.

`pixel_threshold` to 3. We also see that sometimes, a particle momentarily disappears. To still connect these in a
single trace, we want to allow for some gaps in the connection step. Let's use a `window` size of 6 in this test.
Running the algorithm is easy::

>     traces = lk.track_greedy(data, line_width=3, pixel_threshold=3, window=6)

Reviewer: Mention this by name in the text and link to the API reference using `:func:`.

Author: 👍

The result of tracking is a list of kymograph traces::

    >>> print(len(traces))  # the number of traces found in the kymo
    745

Sometimes, we can have very short spurious traces. To remove these from the list of detected traces we can use

> filter_lines. To omit all traces with fewer than 4 detected points, we can invoke::

Reviewer: Link to the API reference whenever possible. It's very convenient for readers.
```suggestion
:func:`filter_lines`. To omit all traces with fewer than 4 detected points, we can invoke::
```

Author: Yeah, I should be more vigilant about opportunities to do so 👍

Reviewer: For some reason it didn't seem to create the API reference link in the generated HTML. Maybe I misremembered the syntax. You may need to fully qualify it: 
```
:func:`~lumicks.pylake.filter_lines`
```

Author: Yeah, that was it!

    >>> traces = lk.filter_lines(traces, 3)

We can get the determined time and coordinate indices of these traces from the `time_idx` and `coordinate_idx`
attributes. Let's plot the first trace::

    plt.plot(traces[0].time_idx, traces[0].coordinate_idx)

Plotting all the detected traces is also quite easy by iterating over the list of traces::

    plt.imshow(data, aspect="auto", vmax=5)
    for trace in traces:
        plt.plot(trace.time_idx, trace.coordinate_idx)

.. image:: kymotracked.png

Fortunately, the signal to noise level in this kymograph is quite good. In practice, when the signal to noise is lower,
one will have to resort to some fine tuning of the algorithm parameters over different regions of the kymograph to get
an acceptable result.

Note that the determined time and coordinates of the traces and coordinates are defined in pixels and therefore need to
be converted to actual time and position using the delta time and pixel size of the kymograph. We can get the time

> between frames directly from the kymograph::
> 
>     dt = kymo.delta_t
> 
> And the pixel size as well::
> 
>     pixel_to_um = kymo.pixelsize_um
> 
> This allows conversion from pixels back to position.

Reviewer: I'm not sure that these need their own lines that just assign to similarly named variables. If they are not used in example code and you just want to mention them, it's better to do it inline and use the form that will link to the API reference:
```suggestion
between frames directly from the kymograph as :attr:`Kymo.delta_t` and the pixel size as :attr:`Kymo.pixelsize_um`. This allows conversion from pixels back to position.
```

Author: Good point.

> The second algorithm present is an algorithm that works purely on signal derivative information. It works by blurring

Reviewer: This should get its own section title since the last one only mentions `greedy`.

```suggestion
Using the lines algorithm
-------------------------

The second algorithm present is an algorithm that works purely on signal derivative information. It works by blurring
```

Author: 👍

but is generally less temporally and spatially accurate due to the blurring involved::

    traces = lk.track_lines(data, line_width=3, max_lines=50)

The interface is mostly the same, aside from an extra required parameter named `max_lines` which indicates the maximum
number of lines we want to detect.

Extracting summed intensities
-----------------------------

Sometimes, it can be desirable to extract pixel intensities in a region around our kymograph trace. We can quite easily
extract these using the method `sample_from_image`. For instance, if we want to sum the pixels in a 9 pixel area around
the longest kymograph trace, we can invoke::

    plt.figure()
    longest_trace_idx = np.argmax([len(trace) for trace in traces])
    longest_trace = traces[longest_trace_idx]
    plt.plot(np.array(longest_trace.time_idx) * kymo.delta_t, longest_trace.sample_from_image(num_pixels=5))
    plt.xlabel('Time [s]')
    plt.ylabel('Summed signal')

Here `num_pixels` is the number of pixels to sum on either side of the trace. Note that data has to be the same image
that the trace was based on.

.. image:: kymo_sumcounts.png

> Algorithms
> ----------
> 

Reviewer: This could be a bit more descriptive. Something along the lines of:
```suggestion
How the algorithms work 
-----------------------

```

Author: 👍

> track_greedy

Reviewer: ```suggestion
:func:`track_greedy`
```

Author: 👍

- Once there are no more peaks in the frame from which we are currently initiating traces, we start initiating traces from the next frame. This process is continued until there are no more peaks left to trace.

The score function is based on a prediction of where we expect future peaks. Based on the peak location of the tip of
the trace `x` and a velocity `v`, it computes a predicted position over time. The score function assumes a Gaussian
uncertainty around that prediction, placing the mean of that uncertainty on the predicted extrapolation. The width of
this uncertainty is given by a base width (provided as sigma) and a growing uncertainty over time given by a diffusion
rate. This results in the following model for the connection score.

.. math::

    S(x, t) = N\left(x + v t, \sigma_{base} + \sigma_{diffusion} \sqrt{t}\right).

Here `N` refers to a normal distribution. In addition to the model, we also have to set a cutoff, after which we deem
peaks to be so unlikely to be connected that they shouldn't be. By default, this cutoff is set at two sigma. Scores
outside this cutoff are set to zero which means they will not be accepted as a new point.

> track_lines

Reviewer: ```suggestion
:func:`track_lines`
```

Author: 👍
